pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+                # operation objects with no owners
+                if self._obs:
+                    raise ValueError(
+                        f"Quantum operation {obj} must occur prior to any measurements."
+                    )
+
+                obj.do_check_domain = False
+
+                # invert the operation if required
+                obj.inverse = info.get("inverse", False)
+
+                if isinstance(obj, STATE_PREP_OPS):
+                    if self._ops:
+                        raise ValueError(
+                            f"State preperation operation {obj} must occur prior to any quantum operations."
+                        )
+
+                    self._prep.append(obj)
+                else:
+                    self._ops.append(obj)
+
+            elif isinstance(obj, MeasurementProcess):
+                # measurement process
+
+                if obj.return_type is qml.operation.Probability:
+                    self._obs.append((obj, obj))
+                    self._output_dim += 2 ** len(obj.wires)
+
+                elif "owns" in info:
+                    # TODO: remove the following line once devices
+                    # have been refactored to no longer use obs.return_type.
+                    # Monkeypatch the observable to have the same return
+                    # type as the measurement process.
+                    info["owns"].return_type = obj.return_type
+                    info["owns"].do_check_domain = False
+
+                    self._obs.append((obj, info["owns"]))
+                    self._output_dim += 1
+
+                    if obj.return_type is qml.operation.Sample:
+                        self.is_sampled = True
+
+            elif isinstance(obj, qml.operation.Observable) and "owner" not in info:
+                raise ValueError(f"Observable {obj} does not have a measurement type specified.")
+
+        self._update()
+
+    def _update_circuit_info(self):
+        """Update circuit metadata"""
+        self.wires = qml.wires.Wires.all_wires(
+            [op.wires for op in self.operations + self.observables]
+        )
+        self.num_wires = len(self.wires)
+
+    def _update_par_info(self):
+        """Update the parameter information dictionary"""
+        param_count = 0
+
+        for obj in self.operations + self.observables:
+
+            for p in range(len(obj.data)):
+                info = self._par_info.get(param_count, {})
+                info.update({"op": obj, "p_idx": p})
+
+                self._par_info[param_count] = info
+                param_count += 1
+
+    def _update_trainable_params(self):
+        """Set the trainable parameters"""
+        self._trainable_params = set(self._par_info)
+
+    def _update(self):
+        """Update all internal tape metadata regarding processed operations and observables"""
+        self._graph = None
+        self._update_circuit_info()
+        self._update_par_info()
+        self._update_trainable_params()
```

Comment:
gradient bookkeeping will also be done here in an upcoming PR

Reply:
Just to get it right, is there a use case where the update function be called several times? Is this to have a circuit, and then change it after?

Reply:
I think I still don't understand...if a tape is nested, and we ask for its wires, does ``tape.wires`` only have the wires visible in the highest level of nesting? And when I expand the tape, could the updating function add new wires that were hidden in a lower level? 

In other words, what metadata does a tape expansion *change*?

Reply:
That's a good question.

I don't think expansion will ever change the wires/number of wires. To work out the wires on a tape, the tape loops through its queue and takes the union of all wires. So this will involve extracting the wires of any nested tapes:

```python
In [5]: with QuantumTape() as tape1:
   ...:     qml.RX(0.5, wires=0)
   ...:     qml.CNOT(wires=[0, 'a'])
   ...:     with QuantumTape() as tape2:
   ...:         qml.Rot(1,1,1,wires=[1])
   ...:         qml.CNOT(wires=[1, 'd'])
   ...:         qml.CNOT(wires=['a', 'd'])
   ...:     qml.PauliX(wires='a')
   ...:

In [6]: tape1.wires
Out[6]: <Wires = [0, 'a', 1, 'd']>

In [7]: tape2.wires
Out[7]: <Wires = [1, 'd', 'a']>
```

Reply:
The reason I've kept these as separate `_update` calls, however, is I can forsee a future quantum transformation that _does_ modify the number of wires (for example, by inserting new operations that act on ancilla wires)